### PR TITLE
add/improve unit tests in FLOPS based mission

### PR DESCRIPTION
**Pull request by xjjiang**
### Summary

This PR adds missing output value checks and partial checks within `mission/flops_based`. Some `test_IO()` functions are added.

Utility function `assert_match_varnames()` is enhanced to allow users to specify `exclude_inputs` and `exclude_outputs`. This allows us to run `test_IO()` even if there are inputs/outputs that are not Aviary variable on purpose.

No unit tests are added for `detailed_landing_phases.py` and `detailed_takeoff_phases.py`. Currently, there is only one case for detailed landing/detailed takeoff. It is currently tested already.

A `TODO` in `mission/flops_based/ode/test/test_mission_eom.py`: This test compares against `full_mission_test_data` in ascent, cruise, and descent phases. The numbers match in ascent and descent phases. My guess is that the number in cruise in `full_mission_test_data` has wrong units. Note: my number 760.55416759 ft/min = 3.86361517135375 m/s which is the expected number.

A `TODO` in `mission/flops_based/ode/test/test_required_thrust.py`: Should `thrust_required` be an Aviary variable? There are other inputs/outputs in `mission/flops_based/ode/` that are not in Aviary variable hierarchy. They can be identified by implementing `test_IO()` function. I didn't do it.

A `TODO` in `mission/flops_based/phases/build_landing.py`: Detailed landing is not currently enabled. This is already a `AnalysisError` in Aviary. It should be copied to [Aviary Issues](https://github.com/OpenMDAO/Aviary/issues).

A `TODO` in `mission/flops_based/phases/build_takeoff.py`: Detailed takeoff is not currently enabled. This is already a `AnalysisError` in Aviary. It should be copied to [Aviary Issues](https://github.com/OpenMDAO/Aviary/issues).

Documentation in `docs/developer_guide/unit_tests.md` is enriched. An interesting statistics:
- `assert_near_equal`: 68%
- `assert_check_partials`: 14%
- `assertEqual`: 8%
- `assert_match_varnames`: 4%
- all other tests: 6%

This PR is part of issue 270 "Complete code coverage for unittesting".

### Related Issues

- Resolves #

### Backwards incompatibilities

None

### New Dependencies

None

### CODE REVIEW

**Comment by xjjiang** [comment]
@jkirk5 I strongly suspect that `ALTITUDE_RATE_MAX` in `validation_cases/validation_data/flops_data/full_mission_test_data.py` should be `760.55416759` instead of `3.86361517135375`. Note that `760.55416759 ft/min = 3.86361517135375 m/s`.

File: aviary/docs/developer_guide/unit_tests.md [hunk 1]
```diff
@@ -27,6 +27,76 @@ Ran 888 tests using 16 processes
...
+```
+
+where the `actual_value` is the value from Aviary and `expected_value` is what the developer expects. Ideally, the `expected_value` should come from computation by another tool (e.g. GASP, FLOPS or LEAPS1) or hand computation. When it is not possible, one can accept an Aviary computed value as expected. This guarantees that future development will not alter the outputs by mistake. As for the tolerance, it is good practice to take 1.e-6. By default, it checks relative error. If the `expected_value` is 0.0, it checks the absolute error.
+
+One can find examples mostly in `subsystems` and `mission` The purpose is to make sure that a variable in an object (namely a component and/or a group) is computed as expected. It is advised that `assert_near_equal` test is carried for all outputs both in components and groups.
+
+### assert_almost_equal
+
+A similar unit test is NumPy's utility is `assert_almost_equal`. It checks whether the absolute difference of `actual_value` from `expected_value` is within certain tolerance. As [documented](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_almost_equal.html) by NumPy, it is not recommented.
```
Comment (Kenneth-T-Moore) [inline aviary/docs/developer_guide/unit_tests.md:46]: recommented -> recommended.

File: aviary/docs/developer_guide/unit_tests.md [hunk 2]
```diff
@@ -27,6 +27,76 @@ Ran 888 tests using 16 processes
...
+
+### assert_check_partials
+
+The second most used assertion is `assert_check_partials` from the OpenMDAO utility. This is critical important because it checks whether the partial derivatives coded by develops are correct. It is the key in optimization. To use this test, you fitst prepare the partial derivative `data` by calling `check_partials` on an object. Then call `assert_check_partials` function with the `data`.
```
Comment (Kenneth-T-Moore) [inline aviary/docs/developer_guide/unit_tests.md:50]: "critical important" -> "critically important"

Comment (xjjiang) [reply aviary/docs/developer_guide/unit_tests.md:50]: done.

File: aviary/docs/developer_guide/unit_tests.md (hunk 1 above)
Comment (xjjiang) [reply aviary/docs/developer_guide/unit_tests.md:46]: done.

File: aviary/docs/developer_guide/unit_tests.md [hunk 3]
```diff
@@ -27,6 +27,76 @@ Ran 888 tests using 16 processes
...
 ```
 
+## Current Unit Tests
+
+### assert_near_equal
+
+The unit test that Aviary uses most is `assert_near_equal` from the OpenMDAO utility [assert_near_equal](https://openmdao.org/newdocs/versions/latest/_srcdocs/packages/utils/assert_utils.html). This assertion takes about 80% of all the assertions. It has the following format:
+
+```
+assert_near_equal(actual_value, expected_value, tolerance=1e-15, tol_type='rel')z
```
[the remark this answers is not in the document]
Comment (xjjiang) [reply aviary/docs/developer_guide/unit_tests.md:37]: done.

File: aviary/mission/flops_based/ode/test/test_mission_eom.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        test on mission EOM using data from validation_cases/validation_data/flops_data/full_mission_test_data.py
+        """
+
+        tol = 1e-6
+        self.prob.run_model()
+
+        # TODO: actual [3679.05255448, 760.55416759, 6557.07891847]
+        #       differ from N3CC from full_mission_test_data
+        #       note: 760.55416759 ft/min = 3.86361517135375 m/s
+        #       My guess is that the full_mission_test_data has a mix of different units.
```
Comment (Kenneth-T-Moore) [inline aviary/mission/flops_based/ode/test/test_mission_eom.py:50]: This is interesting. Looks like the original comes from a LEAPS 1 run. What you concluded seems likely.

**Review by Kenneth-T-Moore — APPROVED** [review]

File: aviary/docs/developer_guide/unit_tests.md [hunk 5]
```diff
@@ -27,6 +27,85 @@ Ran 888 tests using 16 processes
...
+```
+````
+
+### assert_warning
+
+This assertion checks that a warning is issued as expected. Currently, there is only one usage in Aviary but we should add more.
+
+### assert_match_varnames
+
+Another most used assertion is `assert_match_varnames` from Aviary (about 4%). All of them are in `test_IO()` functions. It tests that all of the variables in an object (component or group) that are declared as inputs or outputs exist in the Aviary variable hierarchy. Exceptions are allowed by specifying `exclude_inputs` and `exclude_outputs`. For details, see [assert_utils.py](https://github.com/OpenMDAO/Aviary/blob/main/aviary/utils/test_utils/assert_utils.py).
```
Comment (jkirk5) [inline aviary/docs/developer_guide/unit_tests.md:99]:
```suggestion
Another assertion used in tests is `assert_match_varnames` (about 4%). All of them are in `test_IO()` functions. It tests that all of the variables in an object (component or group) that are declared as inputs or outputs exist in the Aviary variable hierarchy. Exceptions are allowed by specifying `exclude_inputs` and `exclude_outputs`. For details, see [assert_utils.py](https://github.com/OpenMDAO/Aviary/blob/main/aviary/utils/test_utils/assert_utils.py).
```

File: aviary/mission/flops_based/ode/test/test_mission_eom.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        #       3.86361517135375 should be replaced by 760.55416759.
```
Comment (jkirk5) [inline aviary/mission/flops_based/ode/test/test_mission_eom.py:51]:
```suggestion
```

File: aviary/mission/flops_based/ode/test/test_mission_eom.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        # assert_near_equal(prob.get_val(Dynamic.Mission.ALTITUDE_RATE_MAX, units='ft/min'),
+        #                   np.array([3679.0525544843, 3.86361517135375, 6557.07891846677]), tol)
```
Comment (jkirk5) [inline aviary/mission/flops_based/ode/test/test_mission_eom.py:53]:
```suggestion
        assert_near_equal(prob.get_val(Dynamic.Mission.ALTITUDE_RATE_MAX, units='ft/min'),
                          np.array([3679.0525544843, 3.86361517135375, 6557.07891846677]), tol)
```

File: aviary/mission/flops_based/phases/build_landing.py [hunk 8]
```diff
@@ -49,7 +49,7 @@ def build_phase(self, use_detailed=False):
             a group in OpenMDAO
         """
 
-        if use_detailed:
+        if use_detailed:  # TODO
```
Comment (jkirk5) [inline aviary/mission/flops_based/phases/build_landing.py:52]:
```suggestion
        if use_detailed:
```

File: aviary/mission/flops_based/ode/test/test_required_thrust.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        assert_near_equal(
+            self.prob["thrust_required"],
+            np.array([304653.8, 208303.1]), tol
+        )
+
+        partial_data = self.prob.check_partials(out_stream=None, method="cs")
+        assert_check_partials(partial_data, atol=1e-10, rtol=1e-10)
+
+    def test_IO(self):
+        # TODO should thrust_required be an Aviary variable?
```
Comment (jkirk5) [inline aviary/mission/flops_based/ode/test/test_required_thrust.py:52]:
```suggestion
```

**Review by jkirk5 — APPROVED** [review]

File: aviary/docs/developer_guide/unit_tests.md (hunk 5 above)
Comment (xjjiang) [reply aviary/docs/developer_guide/unit_tests.md]: done change.

**Pull request merged**